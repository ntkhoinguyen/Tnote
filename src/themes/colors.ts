const colors = {
  red: "#f03e3e",
  green: "#37b24d",
  blue: "#1c7ed6",
  yellow: "#f59f00",
  black: "#000000",
  white: "#ffffff",
  gray: "#495057",
  pink: "#d6336c",
  violet: "#7048e8",
  indigo: "#4263eb",
  cyan: "#1098ad",
  teal: "#0ca678",
  lime: "#74b816",
  orange: "#f76707",
  dark: "#242424",
  grape: "#ae3ec9",
  light: "#f2f2f2",
  light_gray: "#e6e6e6",
  dark_gray: "#404040",
  light_blue: "#add8e6",
  dark_blue: "#00008b",
  light_green: "#90ee90",
  dark_green: "#006400",
  light_red: "#ff9999",
  dark_red: "#8b0000",
  light_yellow: "#ffffe0",
  dark_yellow: "#808000",
  light_black: "#808080",
  dark_black: "#262626",
};

const colorsPicker = [
  colors.white,
  colors.red,
  colors.green,
  colors.blue,
  colors.yellow,
  colors.cyan,
  colors.pink,
  colors.violet,
  colors.teal,
  colors.lime,
  colors.orange,
  colors.grape,
];

// a,b,c,d,e,f,g,h,i,j,k,l,m,n,o,p,q,r,s,t,u,v,w,x,y,z

const defaultColors = {
  ...colors,
  background: "#ffffff",
  backgroundLoading: colors.white,
  backgroundModal: colors.white,
  backgroundPopup: colors.light,
  border: "#c0c0c0",
  error: "#ff0000",
  headerColor: "#0099FF",
  iconColor: colors.dark,
  iconPopupColor: colors.white,
  inputTextColor: "#000000",
  internetWarning: colors.light,
  kanbanGroupName: colors.light,
  kanbanItem: colors.white,
  geminiBackground: colors.light,
  label: "#000000",
  placeHolderText: colors.light_gray,
  primary: "#0099FF",
  secondary: "#B0B2B4",
  separator: colors.light_gray,
  searchBar: colors.white,
  success: "#00ff00",
  tabbarColor: colors.white,
  tagsHeader: colors.light_gray,
  tagsHeaderText: colors.dark,
  tagsGroup: colors.light_blue,
  text: "#000000",
  userAskBackground: colors.light_blue,
  warning: "#ffff00",
};

const lightColors = {
  ...defaultColors,
};

const darkColors = {
  ...colors,
  // background: "#272924",
  background: "#000000",
  backgroundLoading: colors.gray,
  backgroundModal: colors.dark_black,
  backgroundPopup: colors.dark_black,
  border: "#c0c0c0",
  error: "#ff0000",
  headerColor: colors.black,
  iconColor: colors.white,
  iconPopupColor: colors.dark_black,
  internetWarning: colors.dark,
  inputTextColor: colors.light_gray,
  kanbanGroupName: colors.gray,
  kanbanItem: colors.dark,
  geminiBackground: colors.dark_gray,
  label: colors.white,
  placeHolderText: colors.gray,
  primary: "#0099FF",
  secondary: "yellow",
  separator: colors.white,
  searchBar: colors.dark_black,
  success: "#00ff00",
  tabbarColor: colors.black,
  tagsHeader: colors.dark_black,
  tagsHeaderText: colors.white,
  tagsGroup: colors.light_black,
  text: "#f7f7f7",
  userAskBackground: colors.dark_black,
  warning: "#ffff00",
};

const getColorsByMode = (mode: string) => {
  if (mode === "dark") {
    return { ...lightColors, ...darkColors };
  } else {
    return { ...darkColors, ...lightColors };
  }
};

export {
  lightColors,
  darkColors,
  defaultColors,
  colors,
  colorsPicker,
  getColorsByMode,
};
