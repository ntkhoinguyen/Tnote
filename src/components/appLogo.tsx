// components/AppLogo.tsx
import React from "react";
import { Image, StyleSheet } from "react-native";
import { sizes } from "@/src/themes/sizes";

export const AppLogo = () => (
  <Image
    source={require("@/assets/images/logo.png")}
    style={styles.logo}
    resizeMode="contain"
  />
);

const styles = StyleSheet.create({
  logo: {
    width: sizes.appLogoWidth,
    height: sizes.appLogoHeight,
    marginBottom: sizes.margin.xl,
  },
});

export default AppLogo;
