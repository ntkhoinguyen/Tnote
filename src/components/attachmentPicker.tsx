import React, { useMemo, useState } from "react";
import {
  View,
  Text,
  StyleSheet,
  TouchableOpacity,
  Alert,
  Image,
} from "react-native";
import { CameraView, CameraType, useCameraPermissions } from "expo-camera";
import * as DocumentPicker from "expo-document-picker";

import { useAppContext } from "@/src/useHook/useAppContext";
import { defaultColors } from "@/src/themes/colors";
import { sizes } from "@/src/themes/sizes";
import {
  FontAwesomeIcon,
  MaterialIconsIcon,
  FontAwesome6Icon,
} from "@/src/components/icon";
import {
  AttachmentPickerType,
  OptionImageType,
  RenderCameraType,
} from "@/src/utils/types";
import { useModal } from "@/src/useHook/useModal";

export const AttachmentPicker = (props: AttachmentPickerType) => {
  const { colors, sizes, t } = useAppContext();
  const styles = useMemo(() => createStyles(colors, sizes), [colors, sizes]);

  const { multiple = true, type = "*/*", RenderItem } = props;

  const { open, close, RenderModal } = useModal({
    title: t("selectImage"),
    animationType: "slide",
    containerStyle: { justifyContent: "flex-end" },
    contentStyle: { height: 300 },
    content: (props: any) => (
      <RenderOptionImage onImage={onImage} onFolder={onFolder} />
    ),
  });

  const {
    open: openImage,
    close: closeImage,
    RenderModal: RenderModalImage,
  } = useModal({
    title: t("takeAPhoto"),
    animationType: "slide",
    contentStyle: { height: "100%" },
    content: (props: any) => <RenderCamera onSelect={onSelectImage} />,
  });

  const onPress = () => {
    open();
  };

  const onImage = () => {
    close();
    openImage();
  };

  const pickDocument = async () => {
    try {
      const result = await DocumentPicker.getDocumentAsync({
        type: type,
        multiple: multiple,
        copyToCacheDirectory: true,
      });
      if (result.assets && result.assets.length > 0 && !result.canceled) {
        const files = result.assets.map((asset) => asset.uri);
        props.onSelect(files, false);
      }
    } catch (error) {
      console.log("[attachmentPicker][pickDocument] ----> ", error);
    }
  };

  const onFolder = () => {
    close();
    pickDocument();
  };

  const onSelectImage = (uri: string) => {
    closeImage();
    props.onSelect([uri], true);
  };

  return (
    <TouchableOpacity
      testID="ModalInputSelection"
      onPress={onPress}
      style={styles.container}
    >
      {RenderItem ? (
        <RenderItem />
      ) : (
        <FontAwesomeIcon name="camera" size={14} color={colors.gray} />
      )}

      {RenderModal()}
      {RenderModalImage()}
    </TouchableOpacity>
  );
};

const RenderOptionImage = (props: OptionImageType) => {
  const { colors, sizes, t } = useAppContext();
  const styles = useMemo(() => createStyles(colors, sizes), [colors, sizes]);

  const onImagePress = () => {
    props.onImage();
  };

  const onFolderPress = () => {
    props.onFolder();
  };

  return (
    <View style={styles.OptionsContainer}>
      <TouchableOpacity style={styles.optionItemContent} onPress={onImagePress}>
        <View style={styles.optionItem}>
          <FontAwesomeIcon name="camera" size={40} color={colors.gray} />
        </View>
        <Text style={styles.optionItemText}>{t("camera")}</Text>
      </TouchableOpacity>

      <TouchableOpacity
        style={styles.optionItemContent}
        onPress={onFolderPress}
      >
        <View style={styles.optionItem}>
          <FontAwesomeIcon name="folder" size={40} color={colors.gray} />
        </View>
        <Text style={styles.optionItemText}>{t("folder")}</Text>
      </TouchableOpacity>
    </View>
  );
};

const RenderCamera = (props: RenderCameraType) => {
  const { colors, sizes, t } = useAppContext();
  const styles = useMemo(() => createStyles(colors, sizes), [colors, sizes]);

  const [permission, requestPermission] = useCameraPermissions();

  const [facing, setFacing] = useState<CameraType>("back");
  const [flashMode, setFlashMode] = useState<"on" | "off" | "auto">("off");
  const [uri, setUri] = useState<string | null>(null);

  const cameraRef = React.useRef<CameraView>(null);

  const toggleCameraFacing = () => {
    if (facing === "back") {
      setFlashMode("off");
    }
    setFacing((current) => (current === "back" ? "front" : "back"));
  };

  const toggleFlashMode = () => {
    setFlashMode((current) => (current === "on" ? "off" : "on"));
  };

  const onCancel = () => {
    setUri(null);
  };

  const onSave = () => {
    props.onSelect(uri || "");
  };

  if (!permission) {
    // Camera permissions are still loading.
    return <View />;
  }

  if (!permission.granted) {
    // Camera permissions are not granted yet.
    return (
      <View style={styles.permissionContainer}>
        <Text style={styles.permissionText}>{t("needPermissionCamera")}</Text>
        <TouchableOpacity onPress={requestPermission}>
          <Text style={styles.grantPermission}>{t("grantPermission")}</Text>
        </TouchableOpacity>
      </View>
    );
  }

  const onTakeAPhoto = async () => {
    try {
      if (cameraRef.current) {
        const result = await cameraRef.current?.takePictureAsync?.({
          shutterSound: false,
        });
        if (result.uri) {
          setUri(result.uri);
        } else {
          Alert.alert(t("cannotTakeAPicture"));
        }
      } else {
        Alert.alert(t("cannotTakeAPicture"));
      }
    } catch (error) {
      Alert.alert(t("cannotTakeAPicture"));
      console.error("[camera][onTakeAPhoto] ----", error);
    }
  };

  return (
    <View style={styles.CameraContainer}>
      {uri ? (
        <View style={styles.imageContainer}>
          <Image
            source={{ uri: uri }}
            style={styles.image}
            resizeMode="cover"
          />
          <View style={styles.actionCancelContent}>
            <TouchableOpacity onPress={onCancel}>
              <MaterialIconsIcon name="cancel" size={32} color={colors.gray} />
            </TouchableOpacity>
          </View>
          <View style={styles.actionSaveContent}>
            <TouchableOpacity onPress={onSave}>
              <FontAwesome6Icon
                name="circle-check"
                size={40}
                color={colors.white}
              />
            </TouchableOpacity>
          </View>
        </View>
      ) : (
        <CameraView
          ref={cameraRef}
          style={styles.camera}
          facing={facing}
          flash={flashMode}
          animateShutter={false}
        >
          <View style={styles.buttonContainer}>
            <View style={styles.rightButtons}>
              <TouchableOpacity
                style={styles.button}
                onPress={toggleCameraFacing}
              >
                <MaterialIconsIcon
                  name="flip-camera-ios"
                  size={32}
                  color={colors.white}
                />
              </TouchableOpacity>
              <TouchableOpacity
                style={styles.button}
                onPress={toggleFlashMode}
                disabled={facing === "front"}
              >
                <MaterialIconsIcon
                  name={flashMode === "on" ? "flash-on" : "flash-off"}
                  size={32}
                  color={colors.white}
                />
              </TouchableOpacity>
            </View>

            <TouchableOpacity style={styles.buttonTake} onPress={onTakeAPhoto}>
              <FontAwesomeIcon name="camera" size={32} color={colors.white} />
            </TouchableOpacity>
          </View>
        </CameraView>
      )}
    </View>
  );
};

const createStyles = (colors: typeof defaultColors, size: typeof sizes) => {
  return StyleSheet.create({
    container: {},
    text: {
      fontWeight: size.fontWeight.medium as "500",
    },
    OptionsContainer: {
      flex: 1,
      flexDirection: "row",
      alignItems: "center",
      justifyContent: "space-around",
      paddingVertical: size.padding.sm,
      gap: size.margin.xxl,
    },
    optionItemContent: {
      alignItems: "center",
      justifyContent: "center",
      gap: size.margin.sm,
    },
    optionItem: {
      borderWidth: size.borderWidth.xs,
      borderRadius: size.borderRadius.md,
      borderColor: colors.border,
      padding: size.padding.md,
      alignItems: "center",
      justifyContent: "center",
      backgroundColor: colors.light,
    },
    optionItemText: {
      fontSize: size.fontSize.lg,
      color: colors.text,
    },
    permissionContainer: {
      flex: 1,
      alignItems: "center",
      justifyContent: "center",
      gap: size.margin.lg,
    },
    permissionText: {
      fontSize: size.fontSize.xl,
      color: colors.orange,
    },
    grantPermission: {
      padding: size.padding.md,
      borderWidth: size.borderWidth.xs,
      borderRadius: size.borderRadius.xs,
      borderColor: colors.border,
      backgroundColor: colors.primary,
      color: colors.white,
      fontWeight: size.fontWeight.bold as "bold",
      fontSize: size.fontSize.md,
    },
    CameraContainer: {
      flex: 1,
    },

    camera: {
      flex: 1,
    },
    buttonContainer: {
      flex: 1,
      backgroundColor: "transparent",
      margin: sizes.margin.lg,
      alignItems: "center",
    },
    rightButtons: {
      flex: 1,
      width: "100%",
      alignItems: "flex-end",
    },
    button: {
      marginTop: sizes.margin.xl,
    },
    buttonTake: {
      alignSelf: "center",
      marginBottom: sizes.margin.xxl,
    },
    imageContainer: {
      flex: 1,
    },
    actionImage: {
      flexDirection: "row",
      position: "absolute",
      bottom: 0,
      width: "100%",
      justifyContent: "center",
      gap: sizes.margin.xxl * 3,
    },
    image: {
      height: "100%",
      width: "100%",
    },
    actionSaveContent: {
      position: "absolute",
      width: "100%",
      alignItems: "center",
      bottom: sizes.margin.lg,
    },
    actionCancelContent: {
      position: "absolute",
      width: "100%",
      top: 0,
      alignItems: "flex-end",
      padding: sizes.padding.sm,
    },
  });
};
