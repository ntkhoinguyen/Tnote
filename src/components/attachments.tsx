import React, { useMemo } from "react";
import { View, StyleSheet, Image, Text, TouchableOpacity } from "react-native";

import { useAppContext } from "@/src/useHook/useAppContext";
import { defaultColors } from "@/src/themes/colors";
import { sizes } from "@/src/themes/sizes";
import { AttachmentsType, AttachmentType } from "@/src/utils/types";
import { imageExtensions } from "../utils/utils";
import {
  IoniconsIcon,
  FoundationIcon,
  AntDesignIcon,
  FontAwesomeIcon,
} from "./icon";

export const fileIcons = {
  pdf: (
    <AntDesignIcon
      name="pdffile1"
      size={24}
      color={defaultColors.red}
      style={{ marginTop: sizes.margin.xs }}
    />
  ),
  doc: (
    <FoundationIcon
      name="page-doc"
      size={24}
      color={defaultColors.primary}
      style={{ marginTop: sizes.margin.xs }}
    />
  ),
  docx: (
    <FoundationIcon
      name="page-doc"
      size={24}
      color={defaultColors.primary}
      style={{ marginTop: sizes.margin.xs }}
    />
  ),
  document: (
    <IoniconsIcon
      name="document"
      size={24}
      color={defaultColors.yellow}
      style={{ marginTop: sizes.margin.xs }}
    />
  ),
};

export const Attachments = (props: AttachmentsType) => {
  const { colors, sizes } = useAppContext();
  const styles = useMemo(() => createStyles(colors, sizes), [colors, sizes]);

  const { uri, onPreview, onDelete } = props;

  return (
    <View style={styles.container}>
      {(uri || []).map((item, index) => {
        return (
          <RenderAttachmentItem
            uri={item}
            key={"attachment" + index}
            onPreview={onPreview}
            onDelete={onDelete}
          />
        );
      })}
    </View>
  );
};

const RenderAttachmentItem = (props: AttachmentType) => {
  const { uri, onPreview, onDelete } = props;

  const { colors, sizes } = useAppContext();
  const styles = useMemo(() => createStyles(colors, sizes), [colors, sizes]);

  const fileName = useMemo(() => {
    const split = uri.split("/");
    return split[split.length - 1];
  }, [uri]);

  const fileType = useMemo(() => {
    const split = uri.split(".");
    return split[split.length - 1].toLowerCase();
  }, [uri]);

  const isImage = useMemo(() => {
    return imageExtensions.includes(fileType);
  }, [fileType]);

  const fileIcon = useMemo(() => {
    return (
      fileIcons[fileType as keyof typeof fileIcons] || fileIcons["document"]
    );
  }, [fileType]);

  const onPreviewAttachment = () => {
    onPreview?.(uri);
  };

  const onDeleteAttachment = () => {
    onDelete?.(uri);
  };

  return (
    <TouchableOpacity
      style={styles.attachmentContainer}
      onPress={onPreviewAttachment}
    >
      {isImage ? (
        <Image source={{ uri: uri }} style={styles.image} />
      ) : (
        <View style={styles.IconContent}>
          {fileIcon}
          <View style={{ flex: 1 }}>
            <Text numberOfLines={3} style={styles.text}>
              {fileName}
            </Text>
          </View>
        </View>
      )}
      {onDelete ? (
        <TouchableOpacity style={styles.delete} onPress={onDeleteAttachment}>
          <FontAwesomeIcon name="close" size={24} color={colors.red} />
        </TouchableOpacity>
      ) : null}
    </TouchableOpacity>
  );
};

const createStyles = (colors: typeof defaultColors, size: typeof sizes) => {
  return StyleSheet.create({
    container: {
      flexWrap: "wrap",
      flexDirection: "row",
      width: "100%",
      gap: sizes.margin.sm,
    },
    attachmentContainer: {
      borderRadius: size.borderRadius.sm,
      borderColor: colors.border,
      borderWidth: sizes.borderWidth.xs,
      padding: sizes.padding.sm,
    },
    image: {
      height: 50,
      width: 50,
    },
    IconContent: {
      height: 50,
      width: 150,
      flexDirection: "row",
      backgroundColor: colors.background,
    },
    text: {
      fontSize: size.fontSize.md,
      fontWeight: size.fontWeight.bold as "bold",
      color: colors.text,
      marginLeft: size.margin.sm,
      marginRight: size.margin.lg,
    },
    delete: {
      position: "absolute",
      top: 0,
      right: sizes.margin.xs,
    },
  });
};
