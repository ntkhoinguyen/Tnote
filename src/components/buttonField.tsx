// components/buttonField.tsx
import React, { useMemo, useRef, memo } from "react";
import { TouchableOpacity, Text, StyleSheet } from "react-native";
import { useAppContext } from "@/src/useHook/useAppContext";
import { defaultColors } from "../themes/colors";
import { sizes } from "@/src/themes/sizes";
import { ButtonFieldType } from "../utils/types";
import { getButtonStyles } from "@/src/utils/utils";

export const ButtonField: React.FC<ButtonFieldType> = memo((props) => {
  const { colors, sizes } = useAppContext();
  const {
    text,
    disabled,
    type = "fill",
    color = colors.primary,
    containerStyle = {},
    LeftSection,
    RightSection,
    onPress,
  } = props;

  const styles = useMemo(() => createStyles(colors, sizes), [colors, sizes]);
  const timer = useRef<number | null>(null);
  const { backgroundColor, textColor, borderColor } = useMemo(
    () => getButtonStyles(type, color, colors),
    [type, color, colors]
  );

  const onPressHandler = () => {
    if (timer.current) return;
    if (!disabled && onPress) onPress();

    timer.current = setTimeout(() => {
      timer.current = null;
    }, 2000);
  };

  return (
    <TouchableOpacity
      style={[styles.button, { backgroundColor, borderColor }, containerStyle]}
      disabled={disabled}
      onPress={onPressHandler}
      accessibilityRole="button"
      accessibilityLabel="ButtonField"
    >
      {LeftSection ? <LeftSection /> : null}

      {text ? (
        <Text style={[styles.text, { color: textColor }]}>{text}</Text>
      ) : null}

      {RightSection ? <RightSection /> : null}
    </TouchableOpacity>
  );
});

ButtonField.displayName = "ButtonField";

const createStyles = (colors: typeof defaultColors, size: typeof sizes) =>
  StyleSheet.create({
    button: {
      padding: size.padding.xs,
      borderRadius: size.borderRadius.sm,
      marginVertical: size.margin.sm,
      flexDirection: "row",
      borderWidth: size.borderWidth.xs,
      justifyContent: "space-around",
    },
    text: {
      color: colors.white,
      textAlign: "center",
      fontWeight: size.fontWeight.bold as "bold",
      fontSize: size.fontSize.lg,
      marginHorizontal: size.margin.sm,
    },
  });
