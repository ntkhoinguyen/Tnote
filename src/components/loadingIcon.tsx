import React, { useEffect, useMemo } from "react";
import { View, StyleSheet, Image } from "react-native";
import Animated, {
  Easing,
  useSharedValue,
  useAnimatedStyle,
  withRepeat,
  withTiming,
} from "react-native-reanimated";

import { useAppContext } from "@/src/useHook/useAppContext";
import { defaultColors } from "../themes/colors";
import { sizes, NUM_BARS, indicatorSize } from "@/src/themes/sizes";
import { LoadingIconType } from "../utils/types";

export const LoadingIcon: React.FC<LoadingIconType> = (props) => {
  const { size = "large", color = defaultColors.primary } = props;
  const { colors, sizes } = useAppContext();
  const styles = useMemo(() => createStyles(colors, sizes), [colors, sizes]);
  const indiSize = indicatorSize[size];

  const rotate = useSharedValue(0);

  // ✅ Đặt animation vào useEffect để tránh lỗi
  useEffect(() => {
    rotate.value = withRepeat(
      withTiming(360, {
        duration: 2000,
        easing: Easing.linear,
      }),
      -1,
      false
    );
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);

  const spinnerStyle = useAnimatedStyle(() => ({
    transform: [{ rotate: `${rotate.value}deg` }],
  }));

  return (
    <View
      testID="loadingContainer"
      style={[styles.container, props.containerStyle]}
    >
      <Animated.View
        testID="loadingSpinner"
        style={[styles.spinner, spinnerStyle]}
      >
        {Array.from({ length: NUM_BARS }).map((_, i) => {
          const angle = (360 / NUM_BARS) * i;
          const opacity = (i + 1) / NUM_BARS;

          return (
            <View
              key={i}
              testID={`loadingBar${i}`}
              style={[
                styles.bar,
                {
                  transform: [
                    { rotate: `${angle}deg` },
                    { translateY: indiSize.translateY },
                  ],
                  opacity,
                },
                {
                  width: indiSize.width,
                  height: indiSize.height,
                  backgroundColor: color,
                },
              ]}
            />
          );
        })}
      </Animated.View>
      <Image
        testID="loadingLogo"
        source={require("@/assets/images/logo.png")}
        style={[
          styles.logo,
          { height: indiSize.width * 7, width: indiSize.width * 7 },
        ]}
        resizeMode="contain"
      />
    </View>
  );
};

const createStyles = (colors: typeof defaultColors, size: typeof sizes) => {
  return StyleSheet.create({
    container: {
      backgroundColor: "transparent",
      justifyContent: "center",
      alignItems: "center",
    },
    spinner: {
      position: "absolute",
      justifyContent: "center",
      alignItems: "center",
    },
    bar: {
      position: "absolute",
    },
    logo: {
      zIndex: 1,
    },
  });
};
