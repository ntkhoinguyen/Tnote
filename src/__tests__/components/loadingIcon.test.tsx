import { act, render } from "@testing-library/react-native";

import { LoadingIcon } from "@/src/components/loadingIcon";
import { defaultColors } from "@/src/themes/colors";

describe("LoadingIcon Component", () => {
  beforeEach(() => {
    jest.useFakeTimers();
  });

  afterEach(() => {
    jest.clearAllMocks();
    jest.clearAllTimers();
    jest.useRealTimers();
  });

  test("render init default", () => {
    const { getByTestId, getAllByTestId } = render(
      <LoadingIcon size={"large"} />
    );

    const container = getByTestId("loadingContainer");
    const loadingSpinner = getByTestId("loadingSpinner");
    const bars = getAllByTestId(/^loadingBar/);
    const loadingLogo = getByTestId("loadingLogo");

    expect(container).toBeTruthy();
    expect(loadingSpinner).toBeTruthy();
    expect(bars.length).toBe(13);
    expect(loadingLogo).toBeTruthy();
    expect(loadingLogo.props.source).toBe(require("@/assets/images/logo.png"));
  });

  test("render init small", () => {
    const { getByTestId, getAllByTestId } = render(
      <LoadingIcon size={"small"} />
    );

    const bars = getAllByTestId(/^loadingBar/);
    const logo = getByTestId("loadingLogo");
    expect(bars.length).toBe(13);
    expect(bars[0]).toHaveStyle({
      width: 3,
      height: 7,
      backgroundColor: defaultColors.primary,
    });

    expect(logo).toHaveStyle({
      width: 21,
      height: 21,
    });
  });

  test("render init medium and color pink", () => {
    const { getByTestId, getAllByTestId } = render(
      <LoadingIcon size={"medium"} color={"pink"} />
    );

    const bars = getAllByTestId(/^loadingBar/);
    const logo = getByTestId("loadingLogo");
    expect(bars.length).toBe(13);
    expect(bars[0]).toHaveStyle({
      width: 5,
      height: 17,
      backgroundColor: "pink",
    });

    expect(logo).toHaveStyle({
      width: 35,
      height: 35,
    });
  });

  test("render init large and color red", () => {
    const { getByTestId, getAllByTestId } = render(
      <LoadingIcon size={"large"} color={"red"} />
    );

    const bars = getAllByTestId(/^loadingBar/);
    const logo = getByTestId("loadingLogo");
    expect(bars.length).toBe(13);
    expect(bars[0]).toHaveStyle({
      width: 8,
      height: 20,
      backgroundColor: "red",
    });

    expect(logo).toHaveStyle({
      width: 56,
      height: 56,
    });
  });

  test("check container styles", () => {
    const { getByTestId } = render(
      <LoadingIcon
        size={"large"}
        containerStyle={{ backgroundColor: "yellow" }}
      />
    );
    const container = getByTestId("loadingContainer");
    expect(container).toHaveStyle({
      backgroundColor: "yellow",
    });
  });

  test("Animated test", () => {
    const { getByTestId } = render(<LoadingIcon size={"large"} />);
    const loadingSpinner = getByTestId("loadingSpinner");
    act(() => {
      jest.runAllTimers();
    });
    expect(loadingSpinner).toHaveStyle({
      transform: [{ rotate: "0deg" }],
    });
  });

  test("snapshot test", () => {
    const tree = render(<LoadingIcon size={"medium"} />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
