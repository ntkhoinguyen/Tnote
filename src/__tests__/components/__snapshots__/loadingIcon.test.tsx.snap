// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoadingIcon Component snapshot test 1`] = `
<View
  style={
    [
      {
        "alignItems": "center",
        "backgroundColor": "transparent",
        "justifyContent": "center",
      },
      undefined,
    ]
  }
  testID="loadingContainer"
>
  <View
    style={
      [
        {
          "alignItems": "center",
          "justifyContent": "center",
          "position": "absolute",
        },
        {
          "transform": [
            {
              "rotate": "0deg",
            },
          ],
        },
      ]
    }
    testID="loadingSpinner"
  >
    <View
      style={
        [
          {
            "position": "absolute",
          },
          {
            "opacity": 0.07692307692307693,
            "transform": [
              {
                "rotate": "0deg",
              },
              {
                "translateY": -33,
              },
            ],
          },
          {
            "backgroundColor": "#0099FF",
            "height": 17,
            "width": 5,
          },
        ]
      }
      testID="loadingBar0"
    />
    <View
      style={
        [
          {
            "position": "absolute",
          },
          {
            "opacity": 0.15384615384615385,
            "transform": [
              {
                "rotate": "27.692307692307693deg",
              },
              {
                "translateY": -33,
              },
            ],
          },
          {
            "backgroundColor": "#0099FF",
            "height": 17,
            "width": 5,
          },
        ]
      }
      testID="loadingBar1"
    />
    <View
      style={
        [
          {
            "position": "absolute",
          },
          {
            "opacity": 0.23076923076923078,
            "transform": [
              {
                "rotate": "55.38461538461539deg",
              },
              {
                "translateY": -33,
              },
            ],
          },
          {
            "backgroundColor": "#0099FF",
            "height": 17,
            "width": 5,
          },
        ]
      }
      testID="loadingBar2"
    />
    <View
      style={
        [
          {
            "position": "absolute",
          },
          {
            "opacity": 0.3076923076923077,
            "transform": [
              {
                "rotate": "83.07692307692308deg",
              },
              {
                "translateY": -33,
              },
            ],
          },
          {
            "backgroundColor": "#0099FF",
            "height": 17,
            "width": 5,
          },
        ]
      }
      testID="loadingBar3"
    />
    <View
      style={
        [
          {
            "position": "absolute",
          },
          {
            "opacity": 0.38461538461538464,
            "transform": [
              {
                "rotate": "110.76923076923077deg",
              },
              {
                "translateY": -33,
              },
            ],
          },
          {
            "backgroundColor": "#0099FF",
            "height": 17,
            "width": 5,
          },
        ]
      }
      testID="loadingBar4"
    />
    <View
      style={
        [
          {
            "position": "absolute",
          },
          {
            "opacity": 0.46153846153846156,
            "transform": [
              {
                "rotate": "138.46153846153845deg",
              },
              {
                "translateY": -33,
              },
            ],
          },
          {
            "backgroundColor": "#0099FF",
            "height": 17,
            "width": 5,
          },
        ]
      }
      testID="loadingBar5"
    />
    <View
      style={
        [
          {
            "position": "absolute",
          },
          {
            "opacity": 0.5384615384615384,
            "transform": [
              {
                "rotate": "166.15384615384616deg",
              },
              {
                "translateY": -33,
              },
            ],
          },
          {
            "backgroundColor": "#0099FF",
            "height": 17,
            "width": 5,
          },
        ]
      }
      testID="loadingBar6"
    />
    <View
      style={
        [
          {
            "position": "absolute",
          },
          {
            "opacity": 0.6153846153846154,
            "transform": [
              {
                "rotate": "193.84615384615387deg",
              },
              {
                "translateY": -33,
              },
            ],
          },
          {
            "backgroundColor": "#0099FF",
            "height": 17,
            "width": 5,
          },
        ]
      }
      testID="loadingBar7"
    />
    <View
      style={
        [
          {
            "position": "absolute",
          },
          {
            "opacity": 0.6923076923076923,
            "transform": [
              {
                "rotate": "221.53846153846155deg",
              },
              {
                "translateY": -33,
              },
            ],
          },
          {
            "backgroundColor": "#0099FF",
            "height": 17,
            "width": 5,
          },
        ]
      }
      testID="loadingBar8"
    />
    <View
      style={
        [
          {
            "position": "absolute",
          },
          {
            "opacity": 0.7692307692307693,
            "transform": [
              {
                "rotate": "249.23076923076923deg",
              },
              {
                "translateY": -33,
              },
            ],
          },
          {
            "backgroundColor": "#0099FF",
            "height": 17,
            "width": 5,
          },
        ]
      }
      testID="loadingBar9"
    />
    <View
      style={
        [
          {
            "position": "absolute",
          },
          {
            "opacity": 0.8461538461538461,
            "transform": [
              {
                "rotate": "276.9230769230769deg",
              },
              {
                "translateY": -33,
              },
            ],
          },
          {
            "backgroundColor": "#0099FF",
            "height": 17,
            "width": 5,
          },
        ]
      }
      testID="loadingBar10"
    />
    <View
      style={
        [
          {
            "position": "absolute",
          },
          {
            "opacity": 0.9230769230769231,
            "transform": [
              {
                "rotate": "304.61538461538464deg",
              },
              {
                "translateY": -33,
              },
            ],
          },
          {
            "backgroundColor": "#0099FF",
            "height": 17,
            "width": 5,
          },
        ]
      }
      testID="loadingBar11"
    />
    <View
      style={
        [
          {
            "position": "absolute",
          },
          {
            "opacity": 1,
            "transform": [
              {
                "rotate": "332.3076923076923deg",
              },
              {
                "translateY": -33,
              },
            ],
          },
          {
            "backgroundColor": "#0099FF",
            "height": 17,
            "width": 5,
          },
        ]
      }
      testID="loadingBar12"
    />
  </View>
  <Image
    resizeMode="contain"
    source={
      {
        "testUri": "../../../assets/images/logo.png",
      }
    }
    style={
      [
        {
          "zIndex": 1,
        },
        {
          "height": 35,
          "width": 35,
        },
      ]
    }
    testID="loadingLogo"
  />
</View>
`;
